### PR TITLE
Implement wlr_export_dmabuf_unstable_v1 protocol

### LINKED ISSUE #709: Screenshooting and screencasting
This is just an issue to keep track of prior art and progress of these features in other compositors.

Possible solutions:

* Custom IPC (Sway 0.x)
* D-Bus API (GNOME, KDE) w/ Pipewire
* Wayland protocol (Weston, Orbital - only screenshots)

GNOME and KDE seem to take the [Pipewire](https://pipewire.org/) path with a custom private D-Bus protocol ([here](https://github.com/flatpak/xdg-desktop-portal-gtk/blob/master/src/screencast.c) and [here](https://github.com/KDE/xdg-desktop-portal-kde/blob/screencast/src/screencast.cpp)). They expose a public D-Bus interface via [Flatpak](https://www.flatpak.org/)'s [xdg-desktop-portal](https://github.com/flatpak/xdg-desktop-portal).

If we want to support screenshooters and screencasters using xdg-desktop-portal, we could implement an xdg-desktop-portal backend out-of-tree which talks to the compositor via IPC or a Wayland protocol.

The most efficient way would be to export dmabufs with GBM and give them to the client. That gives a zero-copy, direct access to GPU resources.

References:
* KDE status: http://www.jgrulich.cz/2018/03/06/screen-sharing-in-plasma-wayland-session/
* Private GNOME D-Bus API: https://gitlab.gnome.org/GNOME/mutter/blob/master/src/org.gnome.Mutter.ScreenCast.xml

### PR DESCRIPTION
**Pull request by emersion**
This PR adds support for the [`wlr_export_dmabuf_unstable_v1`](https://github.com/swaywm/wlr-protocols/pull/11) protocol.

- [x] Add dma-buf support to backends
  - [x] On DRM via GBM
  - [x] ~~On other backends via the GL extension (see #906)~~ not possible, that would require to render to a texture
- [x] ~~Support capturing clients~~
- [x] ~~Add byte buffer support to DRM~~
- [x] Implement the protocol
  - [x] Basic functionality
  - [x] Frame synchronization
  - [x] Timestamps
  - [x] ~~Frame locking~~
- [x] Add example client (cc @atomnuker)

Fixes #709

### CODE REVIEW

**Comment by SirCmpwn** [comment]
Also TODO: example client

**Comment by atomnuker** [comment]
Ping? This should be ready for merging.

**Comment by SirCmpwn** [comment]
How am I supposed to use this?

```
~/s/w/build > ./examples/dmabuf-capture 23 vaapi /dev/dri/renderD128 libx264 nv12 12 dmabuf_recording_01.mkv
Capturable output: Dell Inc. Model: DELL P2417H: ID: 21
Capturable output: Goldstar Company Ltd Model: EW224: ID: 22
Capturable output: Goldstar Company Ltd Model: LG Ultra HD: ID: 23
Capturable output: Ancor Communications Inc Model: ASUS VS228: ID: 24
[AVHWDeviceContext @ 0x55e7217dd480] Failed to initialise VAAPI connection: -1 (unknown libva error).
[dmabuf-capture @ 0x7fff63d5cc90] Failed to create a hardware device: Input/output error
```

**Comment by SirCmpwn** [comment]
Troubleshooted here and got further https://wiki.archlinux.org/index.php/Hardware_video_acceleration

<details>

```
~/s/w/build > ./examples/dmabuf-capture 24 vaapi /dev/dri/renderD128 libx264 nv12 12 dmabuf_recording_01.mkv
Capturable output: Dell Inc. Model: DELL P2417H: ID: 21
Capturable output: Goldstar Company Ltd Model: EW224: ID: 22
Capturable output: Goldstar Company Ltd Model: LG Ultra HD: ID: 23
Capturable output: Ancor Communications Inc Model: ASUS VS228: ID: 24
[libx264 @ 0x55cb4fbfe080] using cpu capabilities: MMX2 SSE2Fast SSSE3 SSE4.2 AVX
[libx264 @ 0x55cb4fbfe080] profile High, level 6.1
[libx264 @ 0x55cb4fbfe080] 264 - core 152 r2854 e9a5903 - H.264/MPEG-4 AVC codec - Copyleft 2003-2017 - http://www.videolan.org/x264.html - options: cabac=1 ref=1 deblock=1:0:0 analyse=0x3:0x113 me=hex subme=2 psy=1 psy_rd=1.00:0.00 mixed_ref=0 me_range=16 chroma_me=1 trellis=0 8x8dct=1 cqm=0 deadzone=21,11 fast_pskip=1 chroma_qp_offset=0 threads=12 lookahead_threads=4 sliced_threads=0 nr=0 decimate=1 interlaced=0 bluray_compat=0 constrained_intra=0 bframes=3 b_pyramid=2 b_adapt=1 b_bias=0 direct=1 weightb=1 open_gop=0 weightp=1 keyint=250 keyint_min=25 scenecut=40 intra_refresh=0 rc_lookahead=10 rc=abr mbtree=1 bitrate=12000 ratetol=1.0 qcomp=0.60 qpmin=0 qpmax=69 qpstep=4 ip_ratio=1.40 aq=1:1.00
Output #0, matroska, to 'dmabuf_recording_01.mkv':
    Stream #0:0: Video: h264, nv12, 1920x1080, q=2-31, 12000 kb/s, 60 fps, 1k tbn
[AVHWFramesContext @ 0x55cb4fc01c00] Failed to read image from surface 0x2: 1 (operation failed).
[libx264 @ 0x55cb4fbfe080] AVFrame.format is not set
[libx264 @ 0x55cb4fbfe080] AVFrame.width or height is not set
[libx264 @ 0x55cb4fbfe080] Input picture width (1920) is greater than stride (0)
[dmabuf-capture @ 0x7ffe8d836ec0] Error encoding: Generic error in an external library!
[dmabuf-capture @ 0x7ffe8d836ec0] Wrote trailer!
[libx264 @ 0x55cb4fbfe080] final ratefactor: 42.32
```

</details>

vainfo

<details>

```
libva info: VA-API version 1.1.0
libva info: va_getDriverName() returns 0
libva info: Trying to open /usr/lib/dri/radeonsi_drv_video.so
libva info: Found init function __vaDriverInit_1_1
libva info: va_openDriver() returns 0
vainfo: VA-API version: 1.1 (libva 2.1.0)
vainfo: Driver version: Mesa Gallium driver 18.1.1 for Radeon RX 580 Series (POLARIS10, DRM 3.25.0, 4.17.2-1-ARCH, LLVM 6.0.0)
vainfo: Supported profile and entrypoints
      VAProfileMPEG2Simple            :	VAEntrypointVLD
      VAProfileMPEG2Main              :	VAEntrypointVLD
      VAProfileVC1Simple              :	VAEntrypointVLD
      VAProfileVC1Main                :	VAEntrypointVLD
      VAProfileVC1Advanced            :	VAEntrypointVLD
      VAProfileH264ConstrainedBaseline:	VAEntrypointVLD
      VAProfileH264ConstrainedBaseline:	VAEntrypointEncSlice
      VAProfileH264Main               :	VAEntrypointVLD
      VAProfileH264Main               :	VAEntrypointEncSlice
      VAProfileH264High               :	VAEntrypointVLD
      VAProfileH264High               :	VAEntrypointEncSlice
      VAProfileHEVCMain               :	VAEntrypointVLD
      VAProfileHEVCMain               :	VAEntrypointEncSlice
      VAProfileHEVCMain10             :	VAEntrypointVLD
      VAProfileJPEGBaseline           :	VAEntrypointVLD
      VAProfileNone                   :	VAEntrypointVideoProc
```

</details>

**Comment by atomnuker** [comment]
Either your card doesn't support it, your installation's borked, there's some external libva patches you need, or something I don't know about. The problem's on your end somewhere.

**Comment by SirCmpwn** [comment]
Well, I can't test this until I can narrow it further. Do you know of some good troubleshooting resources? Other vaapi applications which are less experimental I can try?

File: examples/dmabuf-capture.c [hunk 1]
```diff
@@ -0,0 +1,743 @@
+#define _XOPEN_SOURCE 700
```
Comment (agx) [inline examples/dmabuf-capture.c]: Given the recent comments having some usage / trouble shoot  instructions here would IMHO be good

**Comment by SirCmpwn** [comment]
>@acrisci's drm_info will print them out for you

hm, where's this?

**Comment by atomnuker** [comment]
Sorry, I meant @ascent12's drm_info, https://github.com/ascent12/drm_info

**Comment by SirCmpwn** [comment]
This looks good to  Nice work!

**Comment by Sunderland93** [comment]
Does it works with Pipewire?

**Comment by emersion** [comment]
No, but we have plans similar to KDE's to make Pipewire use this. See https://pars.ee/articles/wayland_screen_cap.html

**Comment by emersion** [comment]
(Ping me on IRC if you're interested inb making Pipewire work in sway)

**Pull request merged**